### ISSUE TITLE
Document the new auth system

### ISSUE THREAD

**Issue by timdorr** [timdorr/tesla-api#260]
This is a restart of #215, which has become quite noisy and off topic. **If you are looking for help with a specific language implementation, [please use the Discussion board here on GitHub](https://github.com/timdorr/tesla-api/discussions).**

Tesla has switched to using a separate SSO service (auth.tesla.com) for their app and website. This replaces the previous password grant that was directly on the Owner API. They have also added MFA support with TOTP authenticators. 

I've updated the Ruby library with support for login and token refresh, but need to add MFA support. The docs are completely missing as of this moment. Let's do this!

- [ ] Document new login mechanism
  - [ ] Document MFA path
- [ ] Document refreshing an access token
- [ ] Implement MFA support in Ruby library

**Comment by timdorr** [comment]
I've updated for the new login mechanism: https://tesla-api.timdorr.com/api-basics/authentication

I'm still missing the MFA bits and am working on the refresh token process right now. Please let me know if there are any mistakes. PRs are welcome!

**Comment by Juzzio** [comment]
Hello I would need some help I started to upgrade and while upgrading I have stopped on this :alarm_handler: {:set, {TeslaMate.Vehicles.Vehicle_1_api_error, :fuse_blown}}.   https://pastebin.com/SwCbvJT1

**Comment by alandtse** [comment]
Thanks for documenting this. In my testing of step3, you can't use the base64 encoded version of code_verifier. Are you sure? When I tested with that version I ran into 
```json
{
"error": "invalid_request", 
"error_description": "Invalid code_verifier", 
"error_uri": "https://auth.tesla.com/error/reference/b aabcc7-5c60-43b9-be7c-40a23f7129b1-1612044611898"}
```
I don't have the error if I use the unencoded version.

**Comment by timdorr** [comment]
@alandtse Yep, that's an error on my part. Fixed it up now. Good catch!

**Comment by ekmekcioglu** [comment]
Added JS implementation of oauth2 for v3 with MFA support 
The JS code is in my post  [https://github.com/timdorr/tesla-api/discussions/263](url)

**Comment by Dreamsorcerer** [comment]
If I'm reading the steps correctly, refreshing tokens only works on the initial 5 min tokens? Which presumably requires refreshing within 10 mins of getting the token.

Is there no known way to refresh on a longer term basis? A 10 min refresh period is pretty much unusable for most applications, so we'd end up with the user needing to login again every 45 days, when the access token expires.

**Comment by timdorr** [comment]
There is only a 5 minute time limit on the bearer token, which you should be immediately exchanging for an access token on the Owner API. The refresh token is always long-lived and works until you use it, even after access token expiration.

**Comment by Dreamsorcerer** [comment]
Perfect, thanks.

**Comment by alandtse** [comment]
I'm getting an 401: {"response": "invalid_sso_token"} in step 4 when I use the access_token in step 3.

In step 3, I'm getting an additional id_token:
```json
{"access_token": "eysnip", "refresh_token": "eysnip", "id_token": "eysnip", "expires_in": 300, "state": "snip", "token_type": "Bearer"}
```

Should we be discarding the id_token too? Or can it be used anywhere?

**Comment by timdorr** [comment]
That's for OpenID Connect, which Tesla uses for their website. However, I only see that when refreshing a token. Are you using an authorization_code request_type?

**Comment by alandtse** [comment]
You mean for `grant_type`? Yes.

**Comment by nabilmaad** [comment]
Any idea why in step 2, I keep getting a 200 response with a page not found body instead of a 302 response?

**Comment by alandtse** [comment]
@nabilmaad Is your client following the 302? It will ultimately take you to https://auth.tesla.com/static/404.html. You need to access the last request before that page.

Here's what I'm seeing.
```
POST: redirected from
 https://auth.tesla.com/oauth2/v3/authorize?client_id=ownerapi&code_challenge=SNIP%3D%3D&code_challenge_method=S256&r direct_uri=https://auth.tesla.com/void/callback&response_type=code&scope=openid+email+offline_access&state=fHQ33RtE7BDW1Pie_NhKlKkSCKIqOnlBMsy9XsU_dCE4GImQ_dKBL4M0WpF3Q8NzRtYfdMILi6az4YbDcTT4cQ
GET: redirected from
 https://auth.tesla.com/void/callback?code=SNIP&state=fHQ33RtE7BDW1Pie_NhKlKkSCKIqOnlBMsy9XsU_dCE4GImQ_dKBL4M0WpF3Q8NzRtYfdMILi6az4YbDcTT4cQ&is uer=https://auth.tesla.com/oauth2/v3
GET: https://auth.tesla.com/static/404.html with
```

**Comment by benderl** [comment]
> Any idea why in step 2, I keep getting a 200 response with a page not found body instead of a 302 response?

Maybe the same fix as in my PR to the python example?
https://github.com/enode-engineering/tesla-oauth2/pull/3

**Comment by timdorr** [comment]
@nabilmaad You need to be POSTing, making the OAuth parameters as part of the query string, including all the form fields from the HTML as a form encoded POST body, and include the Cookie header. If you miss any of that, it will return the same HTML page and not the redirect.

**Comment by rustyshelf** [comment]
I had this working, but now at step 1 the site seems to be returning a page with no content and then loading it via Javascript? If I set a custom user agent, I get a page not found HTML response. If I set say Safari's user agent, I get the new javascript page. Anyone else seeing this? Seemed to change about an hour ago.

**Comment by Dreamsorcerer** [comment]
In step 1, it says an 86 character string. I think more precisely, this is 64 bytes of data in base64 encoding.

So, for example, in Python:
```
code_verifier = secrets.token_urlsafe(64)  # 64 bytes
print(len(code_verifier))  # 86 characters
```
Might be worth clarifying that.

**Comment by Dreamsorcerer** [comment]
> The redirect URL. Always "https://auth.tesla.com/void/callback"

I would hope that this is not "Always", and should in fact be anything you want.

This would then allow a standard oauth approach by web applications, so the user never needs to give their credentials to the application. This alternative flow would look like:

1. Create state and other parameters. Redirect the user to that URL.
2. On the callback URL (which the user will be redirected back to), get the passed parameters and verify the state.
3. Continue with step 3/4 as before.

**Comment by minter** [comment]
I'm trying to update my Tesla API app to use the new auth system, and am getting stuck on Step 1. I feel like I'm missing something basic here.

I'm using [Paw](https://paw.cloud) to try to do this, but am also attempting things in the browser (Safari/Mac). I've entered all of the query parameters to send to https://auth.tesla.com/oauth2/v3/authorize in a `GET` as detailed in [Step 1](https://auth.tesla.com/oauth2/v3/authorize).

String values:

```
[20] pry(main)> code_verifier
=> "ZMgDTSwMKQJMDTrVosbkoUxvFYDdEUFcyasxtxPownbOnOUNmzlZBYqHLwxKwlttXxqttIXhVZCKOmPGZZsLwu"
[16] pry(main)> code_challenge = Base64.urlsafe_encode64(Digest::SHA256.hexdigest(code_verifier))
=> "YjFjMDIxNmFiZTRlNGRiOWJjYmUzNzM1MjMwZGMzMTY4MmYwNTVkOTk1N2ExZjZkM2NhMjk4YjlhNGZmZjMwZg=="
[18] pry(main)> code_challenge
=> "YjFjMDIxNmFiZTRlNGRiOWJjYmUzNzM1MjMwZGMzMTY4MmYwNTVkOTk1N2ExZjZkM2NhMjk4YjlhNGZmZjMwZg=="
[19] pry(main)> state
=> "qlQMIsUIyEURnsPQydzkTBIgNDpRnlIpQHOMCKnoemFYBIdySF"
```

<details>
<img width="1280" alt="Screen Shot 2021-01-31 at 12 46 03 PM" src="https://user-images.githubusercontent.com/955/106393094-0fd01600-63c3-11eb-90d4-a0cb7478d4f9.png">
</details>

Those params give me a URL like `/oauth2/v3/authorize?client_id=ownerapi&code_challenge=YjFjMDIxNmFiZTRlNGRiOWJjYmUzNzM1MjMwZGMzMTY4MmYwNTVkOTk1N2ExZjZkM2NhMjk4YjlhNGZmZjMwZg%3D%3D&code_challenge_method=S256&redirect_uri=https:%2F%2Fauth.tesla.com%2Fvoid%2Fcallback&response_type=code&scope=openid%20email%20offline_access&state=qlQMIsUIyEURnsPQydzkTBIgNDpRnlIpQHOMCKnoemFYBIdySF`

When I `GET` that URL either in Paw or Safari, I get redirected to a 404 page.

If I go directly to https://auth.tesla.com/ in Safari, I get redirected to this URL: https://auth.tesla.com/oauth2/v1/authorize?client_id=teslaweb&response_type=code&scope=openid%20email%20profile&redirect_uri=https%3A//www.tesla.com/openid-connect/generic&state=aY0etlKMmsia795qdg52tZ-w2zkup90_XmDeCBj4VV0&locale=en-US

That's `v1` OAuth, not a `v3`.

Can anyone see what mistake I'm making here to even pull the login screen up?

**Comment by baysinger** [comment]
@minter I don't think you're doing anything wrong, other than expecting the redirect to work. The redirect URL is invalid, but the Tesla app intercepts the Location header and pulls the code out of it. So if you want to paste the URL into a web browser, you'll need one with good developer tools, so you can see the request for https://auth.tesla.com/void/callback and copy the code out of the URL.

Here's what I see in Firefox's network tool:

![image](https://user-images.githubusercontent.com/16112955/106396228-8cf78d00-63bb-11eb-854d-640081a79176.png)

**Comment by baysinger** [comment]
@minter That URL takes me to the Telsa auth page, which requests username and password. But it you're already logged in, it might be taking you directly to the callback, which is an invalid URL. (The callback is supposed to be invalid. And AFAIK, you can't change it to anything else.)

If I'm already logged in, I get this flow:
![image](https://user-images.githubusercontent.com/16112955/106396425-ab11bd00-63bc-11eb-888b-3a29e2d9547e.png)

So that might match what's happening for you. The 404 is totally normal.

**Comment by borgstrom** [comment]
Hi @timdorr 👋 

Your comment here https://github.com/timdorr/tesla-api/issues/215#issuecomment-770139734 is incorrect.

The server does not discard `state`. It gives it back to you when you exchange your challenge code for a a bearer token.

You're correct that it is just recommended and not required as part of the spec. I also understand that right now the docs don't deal with MFA, and as such it means that the user is not actually involved with the auth flow right now so hijacking the flow is unlikely. However, once MFA is involved it means that consumers of this documentation will need to have a step involved where they prompt the user for an MFA code and maintain state themselves, at which point doing the state validation to prevent CSRF becomes much more critical.

I believe that the documentation should encourage users to validate the state they get back from exchanging the challenge code for the bearer token matches what they generated in all circumstances. Guidance should be to simply fail if the state does not match. For HTTP servers returning a 400 would be appropriate. For other use cases they can simply abort the auth process.

**Comment by nabilmaad** [comment]
I've made some progress in step 2 where I'm now seeing a 302 response with a `Location` header of `https://auth.tesla.com/oauth2/v3/authorize?client_id=ownerapi&code_challenge=<my_code_challenge>&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fauth.tesla.com%2Fvoid%2Fcallback&response_type=code&scope=openid+email+offline_access&state=<my_state>`. This is followed by a GET to the that URL which is giving my the login page again (with a 200). Looks like I'm missing a request that has the `code` param in it like @alandtse and @baysinger are seeing.

I suspect I'm doing something wrong with either setting the cookie for the POST, or the body itself. I'm always getting 2 cookies from the first request, one that always starts with `tesla-auth.sid=` and the other always starts with `TS`. Which one of the 2 should I use? Should I copy the entire thing as is or just the value itself (i.e. ignoring the cookie name and everything that comes after the ;)? I've tried multiple combinations with no luck.

Here's a screenshot of how I'm setting the POST body in case someone sees anything that looks off. I set Paw to `Follow Redirects` in the request options.


<img width="721" alt="image" src="https://user-images.githubusercontent.com/2532658/106396897-be328600-63d8-11eb-9e52-6933705675f5.png">

**Comment by alandtse** [comment]
@Dreamsorcerer 
> I would hope that this is not "Always", and should in fact be anything you want.
> 

> This would then allow a standard oauth approach by web applications, so the user never needs to give their credentials to the application. This alternative flow would look like:

I don't think you can change it. I'm testing whether I can change the redirect_url and it looks like Tesla detects this.
![image](https://user-images.githubusercontent.com/7086117/106397245-31c89900-63c1-11eb-9851-3d67891625a5.png)

I'd be happy to know I'm wrong though.

**Comment by timdorr** [comment]
> I would hope that this is not "Always", and should in fact be anything you want.
> 
> This would then allow a standard oauth approach by web applications, so the user never needs to give their credentials to the application. This alternative flow would look like:

It cannot. That is a known attack vector for OAuth. You must lock it down to _at least_ a particular domain, if not a specific URL. Otherwise, that opens you up to attack and account takeover.

**Comment by timdorr** [comment]
Folks, I'll ask again: **If you need help with implementing the API or auth system, [please use the Discussion board here on GitHub](https://github.com/timdorr/tesla-api/discussions).** 

I'll be hiding any requests for help that already exist and will delete any such comments from here on out. This issue is for discussion around the API, it's documentation, and the Ruby library within this repo. This is not a support system, so please keep those kinds of requests in [the Discussions tab](https://github.com/timdorr/tesla-api/discussions).

**Comment by timdorr** [comment]
> The server does not discard `state`. It gives it back to you when you exchange your challenge code for a a bearer token.

Yes, that is technically correct. You should normally validate your `state` between steps of the authorization flow to prevent some MITM attacks. However, in this case, since we have full control over the user agent (because we _are_ the user agent), it's not strictly necessary. The kinds of attacks it prevents are not possible. 

Hence, discarding it is safe and serves to reduce the complexity in implementing this auth flow. But if folks want to get it exactly to spec, that's certainly an alright thing to do.

**Comment by baysinger** [comment]
@timdorr I believe the base64 encoding has to be stripped of its padding (equals sign). There's an example of that in the Python discussion, and and I had to do that to my script to get it working. I don't see mention of that in your documentation.

**Comment by timdorr** [comment]
Yes, that's what's meant by URL-safe base64 encoding. Here's the spec: https://tools.ietf.org/html/rfc4648#section-5

**Comment by baysinger** [comment]
I think some more clarity is needed. By default, the URL-safe base64 function in Ruby (and in Python) still do that padding. Looking at the Ruby documentation (https://ruby-doc.org/stdlib-2.4.0/libdoc/base64/rdoc/Base64.html), the padding parameter is implicitly set to true. So you'd have to explicitly pass false for that parameter, which you're not doing in the example. (Perhaps there a different version of the Base64 library where it's implicitly false?)

Also, the RFC you point to essentially says the padding is optional, whereas to make the Tesla authentication work, it seems padding is disallowed. So I think that should be called out in the documentation.

**Comment by timdorr** [comment]
Yeah, sorry, the padding is allowed by Tesla's API. You can see it in this successful login recording in the Ruby tests: 

https://github.com/timdorr/tesla-api/blob/74726ce5cb5b0f47cfca1c2f481e019133de78a3/spec/cassettes/client-login.yml#L5

**Comment by baysinger** [comment]
Ok, my mistake.  Where I do have to strip the padding is for code_verifier, if I use base64 encoding, as @Dreamsorcerer says above. In your log, you are using a code_verifier that resembles base64 without padding, but in your code, you are not using a base64 encoding function:
`code_verifier = rand(36**86).to_s(36)`

While that works, you can't just use any string for code_verifier. You have to use a string that resembles a url-safe base64 encoding of a 64-element byte array, without padding, which comes out to an 86-character string. Or a least, in my testing, that appears to be the case.

**Comment by timdorr** [comment]
I added some clarification that it needs to be alphanumeric. That should be sufficient, as the main point is that it can go into a URI without further encoding. More specifically, [it needs to use only the unreserved characters](https://tools.ietf.org/html/rfc7636#section-4.1) from [the URI spec](https://tools.ietf.org/html/rfc3986#section-2.3). It's common to base64url or base36 the value, since that also gets you to the same kind of string.

I haven't actually tried other lengths allowed by the spec (43-128). But just to be clear, it is represented by character length, not byte length, in the spec. I have no idea how much of this is bespoke implementation by Tesla or if they're using an off the shelf solution that has certain behavior semantics (including ones that are outside of the spec).

**Comment by Dreamsorcerer** [comment]
> It cannot. That is a known attack vector for OAuth. You must lock it down to _at least_ a particular domain, if not a specific URL. Otherwise, that opens you up to attack and account takeover.

How else would you support authentication in 3rd-party apps? e.g. How you can login to a website via OpenID, Facebook or Github.

**Comment by Dreamsorcerer** [comment]
> But just to be clear, it is represented by character length, not byte length, in the spec. I have no idea how much of this is bespoke implementation by Tesla or if they're using an off the shelf solution that has certain behavior semantics (including ones that are outside of the spec).

Yes, but the bit that is not obvious is how it _should_ be created (rather than what is the bare minimum requirement), as described in the spec you linked:

> It is RECOMMENDED that the output of a suitable random number generator be used to create a 32-octet sequence.  The octet sequence is then base64url-encoded to produce a 43-octet URL safe string to use as the code verifier.

Therefore, the recommendation would be to use (in Python) `secrets.token_urlsafe(64)` rather than some other method of generating 86 random characters.

**Comment by baysinger** [comment]
> 
> 
> > It cannot. That is a known attack vector for OAuth. You must lock it down to _at least_ a particular domain, if not a specific URL. Otherwise, that opens you up to attack and account takeover.
> 
> How else would you support authentication in 3rd-party apps? e.g. How you can login to a website via OpenID, Facebook or Github.

I'm guessing the redirect URI is locked to the client ID. So if Tesla ever supports third party apps, you would create a new client profile which contains your URI, and then it would work for your app. But since we're stuck using the client ID from Tesla's mobile app, we're also stuck with their redirect URI.

**Comment by timdorr** [comment]
> How else would you support authentication in 3rd-party apps? e.g. How you can login to a website via OpenID, Facebook or Github.

You provide the redirect URL to the provider ahead of time. Almost all OAuth providers require you to set a redirect URL when creating the client ID/secret for your application. 

Here's Twitter's docs about it: https://developer.twitter.com/en/docs/apps/callback-urls

**Comment by jasminsehic** [comment]
> I haven't actually tried other lengths allowed by the spec (43-128). But just to be clear, it is represented by character length, not byte length, in the spec. I have no idea how much of this is bespoke implementation by Tesla or if they're using an off the shelf solution that has certain behavior semantics (including ones that are outside of the spec).

Does anyone know where the 86 character limit came from? Curios to confirm this. OAuth spec says:
![image](https://user-images.githubusercontent.com/3346727/106413926-458a0300-6486-11eb-9bd1-3044b1f7bc57.png)

**Comment by Dreamsorcerer** [comment]
> > I haven't actually tried other lengths allowed by the spec (43-128). But just to be clear, it is represented by character length, not byte length, in the spec. I have no idea how much of this is bespoke implementation by Tesla or if they're using an off the shelf solution that has certain behavior semantics (including ones that are outside of the spec).
> 
> Does anyone know where the 86 character limit came from? Curios to confirm this. OAuth spec says the same (43-128 range) and that it can be base66 (where URL safe base64 and alphanumeric would fit into anyway):

I'm assuming the usage was just extracted from the Tesla app, and appeared to be passing an 86 character string (as everything we know is just reverse-engineered). As I mentioned previously, the implementation is most likely generating 64 bytes of secure random data and then base64 encoding it.

**Comment by jasminsehic** [comment]
> > > I haven't actually tried other lengths allowed by the spec (43-128). But just to be clear, it is represented by character length, not byte length, in the spec. I have no idea how much of this is bespoke implementation by Tesla or if they're using an off the shelf solution that has certain behavior semantics (including ones that are outside of the spec).
> > 
> > 
> > Does anyone know where the 86 character limit came from? Curios to confirm this. OAuth spec says the same (43-128 range) and that it can be base66 (where URL safe base64 and alphanumeric would fit into anyway):
> 
> I'm assuming the usage was just extracted from the Tesla app, and appeared to be passing an 86 character string (as everything we know is just reverse-engineered). As I mentioned previously, the implementation is most likely generating 64 bytes of secure random data and then base64 encoding it.

I can confirm that using base66 and various lengths doesn't seem to work. I get internal server error on token exchange so I assume there is something non-standard on their side. Not sure if it is the length or character set but I am going to stick to 86 alphanumeric for now as that seems to work consistently.

**Comment by timdorr** [comment]
I've finished [adding MFA support to the Ruby gem](https://github.com/timdorr/tesla-api/blob/49d1ccedcceaf765099f50df68e636385a78be72/lib/tesla_api/client.rb#L101-L138), so that should be good to go. 

Just to summarize that code, you have to look at the response HTML from the form POST and see if it looks like an MFA form page (I'm searching for `passcode` in the body). From there, you use the `transaction_id` from the original page to obtain a list of MFA factors (devices/apps), and verify the passcode for the transaction. Finally, you finish up by submitting the form once again, but with only the `transaction_id` parameter in the body. Afterwards, you continue as normal.

I'll try to get that in the docs soon, but I've got the code up to look over for now.

**Comment by fredli74** [comment]
I'm pretty sure you do not need `client_secret` when doing `refresh_token` on the new `/oauth2/v3/token` end-point. As far as I know it's only the old owner-api that needs the `client_secret`.

You do however need the `scope` as you mentioned in your documentation. The refresh on `/oauth2/v3/token` works without it, and everything looks fine, but when you try to use the `access_token` to access `/oauth/token` you will *then* get a 401 error telling you unknown user. That kinda threw me off for a bit until I realized what I had missed.

**Comment by jasminsehic** [comment]
> I'm pretty sure you do not need `client_secret` when doing `refresh_token` on the new `/oauth2/v3/token` end-point. As far as I know it's only the old owner-api that needs the `client_secret`.

Yep `client_secret` is not required to refresh token on oauth2

**Comment by timdorr** [comment]
Fixed that up now.

**Comment by philhzss** [comment]
I used to use Postman to test stuff with the Tesla API before writing it into my little projects, but I just reviewed the documentation you're working on and I think that's no longer possible right? 
(No longer possible to do a few GET and POSTs in Postman to test the API, since it seems that it requires much more code to authorize)
Thank you!

**Comment by qJake** [comment]
Has anyone tested to see if Tesla's OpenID implementation supports [Dynamic Client Registration](https://openid.net/specs/openid-connect-registration-1_0.html)? This would essentially mean that third-parties could register for their own Client ID (and specify their own set of Redirect URIs) while still complying with the spec and allowing Tesla to retain control.

**Comment by alandtse** [comment]
> Has anyone tested to see if Tesla's OpenID implementation supports [Dynamic Client Registration](https://openid.net/specs/openid-connect-registration-1_0.html)?

I played with the https://auth.tesla.com/oauth2/v3/register endpoint and it does not appear to recognize it. Posting to the endpoint would result in an error saying the response_type was missing. You could avoid the error by putting the response_type into the url, but it's not apparently responding the way the OpenId spec suggests it should so I stopped experimenting. I'm no expert so perhaps someone else can figure it out.

Of course this may be the wrong endpoint but was my first guess.

**Comment by lannsjo** [comment]
Big thanks to everyone for this, I have now a working solution in C# using the basics of this [python code ](https://github.com/enode-engineering/tesla-oauth2) running as a lambda in a serverless AWS environment. 
Authentication with and without mfa works. 
Refreshing token works as wll. Yeay!

**Comment by timdorr** [comment]
@alandtse It may also be under the `/oauth2/v1/` namespace. They use that for at least two different client IDs: `teslaweb` and `mfaportal`. 

No idea what happened to `v2` 🤷

**Comment by mseminatore** [comment]
As of today I am finding that the hidden form fields are missing. Is there a new change to the new auth flow?

**Comment by fredli74** [comment]
> As of today I am finding that the hidden form fields are missing. Is there a new change to the new auth flow?
Nope, still have them.
```
            <input type="hidden" name="_csrf" value="..." />
            <input type="hidden" name="_phase" value="authenticate" />
            <input type="hidden" name="_process" value="1" />
            <input type="hidden" name="transaction_id" value="..." />
            <input type="hidden" name="cancel" value="" id="form-input-cancel" />```

**Comment by mseminatore** [comment]
Hmm. Then maybe I've messed up something since yesterday. Can anyone tell me if this looks valid for a code_challenge? I assumed that challenge is UTF-8. Is 43 characters correct for a base64 encode?

fejEQANXrR8IcHuMPqgamH0U9gPvL9JH2t3YA96G54g

**Comment by mseminatore** [comment]
Strange, now it appears to be working again. I think Tesla must be messing around with auth flow or having some instabilities.

**Comment by jasminsehic** [comment]
I think there is a WAF in place where if you make too many requests frequently with Tesla SSO you get directed to an empty page that waits a little while before redirecting you.

**Comment by mseminatore** [comment]
Ah. So perhaps in debugging my code I tripped a throttling limit. I'll let it cool down for a while and then try again.

**Comment by jasminsehic** [comment]
I've stuck with hashing an 86 character length alphanumeric string into an URL safe Base64 string. Anything outside of this spec I've run into some sort of issues.

**Comment by mseminatore** [comment]
Any ideas what timeframe the WAF imposes? I want to play nice and stay under the limit.

**Comment by mseminatore** [comment]
@fredli74 Could you (or someone else) share the top part of the result body with the hidden input forms? In HTML form this is what I am receiving:

`Please enable JavaScript to view the page content.
Your support ID is: 10100662323809908378.`

**Comment by rustyshelf** [comment]
@mseminatore I had that too, it generally means you've successfully logged in and now have a login cookie in your session. If you clear it out you should be back to square 1.

**Comment by mseminatore** [comment]
Interesting. I did just complete a successful login using TeslaJS so that would make sense. I am now trying to debug the Tesla Control app where I need to fix the login flow to match the new one. Code is very similar but not identical (NodeJS vs. UWP). I don't know where a UWP app would store a cookie.

**Comment by jasminsehic** [comment]
> Any ideas what timeframe the WAF imposes? I want to play nice and stay under the limit.

Not sure, I would say 15-30 minutes. Whenever I've got stuck with that `Please enable JavaScript` response I had to give up and come back later. Few hours later it worked. It had nothing to do with cookie for me as my solution is stateless between auth attempts so this is very likely a WAF on Tesla end while we are debugging we are hitting that endpoint quite a bit.

**Comment by edlea** [comment]
> I think there is a WAF in place where if you make too many requests frequently with Tesla SSO you get directed to an empty page that waits a little while before redirecting you.

That might well be what it is. I've definitely [been seeing](https://github.com/timdorr/tesla-api/discussions/279) behaviour that matches that.

After some trial and error, it seems that retrying after a bit of a pause will eventually (where "eventually" < 30s) get a working page. Looks like a [common workaround](https://github.com/enode-engineering/tesla-oauth2/blob/4799c957d8bdc5e983d7bfaaf48d80f3d5754044/tesla.py#L33) for now.

I'd be surprised if this is what the official Tesla app is doing though.

**Comment by jasminsehic** [comment]
> > I think there is a WAF in place where if you make too many requests frequently with Tesla SSO you get directed to an empty page that waits a little while before redirecting you.
> 
> That might well be what it is. I've definitely [been seeing](https://github.com/timdorr/tesla-api/discussions/279) behaviour that matches that.
> 
> After some trial and error, it seems that retrying after a bit of a pause will eventually (where "eventually" < 30s) get a working page. Looks like a [common workaround](https://github.com/enode-engineering/tesla-oauth2/blob/4799c957d8bdc5e983d7bfaaf48d80f3d5754044/tesla.py#L33) for now.
> 
> I'd be surprised if this is what the official Tesla app is doing though.

With actual browser implementations it is not an issue because the page loaded by WAF has javascript execute that sets an additional cookie from client side which then bypasses the WAF and goes to the actual auth page. Any solution where we are scraping HTML using requests without a javascript engine is vulnerable to Tesla forcing this rule by default. I can’t really see a good way to get around this other than to proxy requests through a remote web driver etc but hopefully Tesla doesn’t pull the plug on us all and they allow us to officially apply to be an authorized Tesla integrated app before they do that.

**Comment by corsair** [comment]
An important notion regarding the "common workaround" of hitting the OAuth pages multiple times to work around the WAF is that if you've reached the "POST" portion (Step #2 in @timdorr's instructions) and it takes a few tries, it increases the risks of triggering an account lockout and requiring the account holder to reset their password. This same issue also applies to providing the wrong passcode multiple times.

I would **highly suggest** that any libraries that use the scraping method try to immediately return an error back upon determining the credentials are incorrect instead of allowing the subsequent attempts to go through. Granted, as shared by @jasminsehic above, there isn't always a good solution to working around this fully.

This risk increases with the more third-party services/apps that an individual Tesla Account utilizes, so it should be our responsibility to try to tune this as much as possible. I've reduced my "tries" to a maximum of three as I've received a few emails within the last few days from members having their accounts locked due to a combination of my own service and another causing lockouts.

**Comment by edlea** [comment]
> An important notion regarding the "common workaround" of hitting the OAuth pages multiple times to work around the WAF is that if you've reached the "POST" portion (Step # 2 in @timdorr's [instructions](https://tesla-api.timdorr.com/api-basics/authentication)) and it takes a few tries, it increases the risks of triggering an account lockout and requiring the account holder to reset their password. This same issue also applies to providing the wrong passcode multiple times.

Very good point. To be honest, I hadn't read further down the code I linked to and had assumed the retries are only for the first GET request for the HTML; that's the implementation route I've taken. Other requests (with valid cookies?) don't seem to have the same issue.

**Comment by corsair** [comment]
> > An important notion regarding the "common workaround" of hitting the OAuth pages multiple times to work around the WAF is that if you've reached the "POST" portion (Step # 2 in @timdorr's [instructions](https://tesla-api.timdorr.com/api-basics/authentication)) and it takes a few tries, it increases the risks of triggering an account lockout and requiring the account holder to reset their password. This same issue also applies to providing the wrong passcode multiple times.
> 
> Very good point. To be honest, I hadn't read further down the code I linked to and had assumed the retries are only for the first GET request for the HTML; that's the implementation route I've taken. Other requests (with valid cookies?) don't seem to have the same issue.

That makes sense regarding the cookies! I've made an [issue](https://github.com/enode-engineering/tesla-oauth2/issues/8) in the aforementioned repository to recommend not utilizing multiple attempts (or otherwise greatly reduce them) if we've gotten past the initial request of the initial form. I would recommend other implementations handle this similarly until more information is available.

**Comment by jasminsehic** [comment]
I've had the `/oauth2/v3/authorize/mfa/factors` return the `Please enable Javascript` page on me a few times and that has cookies on it too so looks like the WAF doesn't care the session cookie is there, probably whether the other client side created cookie is there or not which I expect is somehow digitally signed. Javascript is heavily obfuscated, not worth trying to reverse.

**Comment by mrgaolei** [comment]
When I POST `code`, `code_verifier` and 3 other params to `https://auth.tesla.com/oauth2/v3/token`

returns 403 error with body:
```json
{"error":"invalid_grant","error_description":"Internal server error","error_uri":"https://auth.tesla.com/error/reference/2a7ba50b-f676-4851-a15e-b63c4cccdfa0-1612365515686"}
```

which step I wrong?

**Comment by SearchForTheCode** [comment]
So i am currentlz having trouble with implementing the new flow!!
In the Java lib the NoMFA way does not work with 
https://github.com/tbee/teslaApi/blob/main/src/main/java/org/tbee/tesla/TeslaNoMFALogic.java with "401 "unsupported SSO app" 

I guess there is no solution to this yet. The MFA flow https://github.com/tbee/teslaApi/blob/main/src/main/java/org/tbee/tesla/TeslaMFALogic.java 
should work (from commit comments) but i couldnt test it yet. Before i destroy my work, i just want to clarify if there is any solution to the unsupported SSO app .

**Comment by Dreamsorcerer** [comment]
> If you are looking for help with an implementation or library/tool that uses the API, [please use the Discussion board here on GitHub](https://github.com/timdorr/tesla-api/discussions).

**Comment by timdorr** [comment]
I added a note about the WAF stuff at the top of the auth docs to hopefully avoid some pain points.

**Comment by mseminatore** [comment]
I figured out my issue with no form. I was improperly formatting my request URL due to differences between NodeJS and WinJS. Having carefully implemented a WinJS equivalent of the NodeJS request library I've been able to get much further in the process. Taking care to keep the number of attempts low to be nice to the servers.

I am hoping to get Tesla Control back working again shortly before everyone's current tokens expire! :)

**Comment by mseminatore** [comment]
OK, now I am at the point where I am failing to retrieve a location header. I suspect there is a bug in my POST of the form.

Does the following look right for the POST body payload for the form submission?

`_csrf=cEGPYxk6-tZYFq-hnKYQuGgdObBdkCWm01Mw&_phase=authenticate&_process=1&transaction_id=HSnCC7Md&cancel=&identity=myemail&credential=mypassword`

**Comment by mrgaolei** [comment]
In step 3, I post:
```python
resp = requests.post("https://auth.tesla.com/oauth2/v3/token", json={
            'grant_type': 'authorization_code',
            'client_id': 'ownerapi',
            'code': code,
            'code_verifier': code_verifier,
            'redirect_uri': 'https://auth.tesla.com/void/callback'
        })
```
but returns HTTP 403:
```json
{"error":"invalid_grant","error_description":"Internal server error","error_uri":"https://auth.tesla.com/error/reference/2701fe84-39c3-4ae8-b771-84ae6fbb7550-1612377093226"}
```

I tried many code_verifier and your test case: `dxqer47aeddb0wgk7rt99sk58y05dh0ylbxf2hqfr72o44gi41e6xw4wro6kg3sjx0xr2kythlvcqea1doz6pf`
What should I do?

**Comment by mseminatore** [comment]
I noticed that I am not seeing a Set-Cookie header returned in the response from Step 1. Is that expected? I am using XMLHttpRequest(), that should see the Set-Cookie header if it is returned right?

**Comment by jasminsehic** [comment]
> I noticed that I am not seeing a Set-Cookie header returned in the response from Step 1. Is that expected? I am using XMLHttpRequest(), that should see the Set-Cookie header if it is returned right?

XMLHttpRequest() strips out Set-Cookie header for security reasons. Depending on what your environment is you need to use something else. If it is NodeJS use `axios` or `node-fetch`. If you are using a real browser you should look at popping up the auth screen if you can so user can interact with it directly.

**Comment by mseminatore** [comment]
This is an HTML/JS Universal Windows App. If a cookie is being sent, the NodeJS Request library seems to be handling it properly.

It would be very unfortunate if the cookie is required and there is no way for me to do this from an HTML/JS app. Hoping someone has ideas.

**Comment by timdorr** [comment]
The Cookie header is a hard requirement for the POST request to succeed, unfortunately.

**Comment by edlea** [comment]
I don’t know what’s possible in a Universal Windows App, but if you can pop up a browser (Edge?) window and open the authorise URL in step 1 and then watch for, and catch, the “void” callback URL with the code you might be okay?

I’m considering this route in an app and my limited testing suggests user-agent isn’t that important so might work.

**Comment by lannsjo** [comment]
I have a lot of users that still have the old token and refreshtoken in my system. 

The refreshtoken will soon expire for most of my users. As I understand it, I will not be able to refresh their tokens with the new auth-model.. They will be forced to login again to be able to use the service.. am I right? Would really like to not forcing them to login again.

**Comment by corsair** [comment]
> I have a lot of users that still have the old token and refreshtoken in my system.
> 
> The refreshtoken will soon expire for most of my users. As I understand it, I will not be able to refresh their tokens with the new auth-model.. They will be forced to login again to be able to use the service.. am I right? Would really like to not forcing them to login again.

There is, unfortunately, no way of avoiding this. The only way of avoiding this is if you retained their credentials (which is a very bad practice / not recommended) and even then, this wouldn't work if an account has 2FA enabled.

The best route is to inform your users in some manner, whether it is updating your "invalid token" notifications/emails to inform them of these changes being the causation and hope for the best.

**Comment by mseminatore** [comment]
I don't know that I have the ability to intercept the callback. Wouldn't
that involve injecting my own JS code into the browser window
containing the web browser? I'll have to look into it because otherwise I
may be finished and have to re-write the entire app in C#.

Anyone know if XMLHttpRequest throws away cookies or will it capture and
send them on subsequent calls? I don't think there is a need to manipulate
them or access them directly.

**Comment by jasminsehic** [comment]
> Anyone know if XMLHttpRequest throws away cookies or will it capture and send them on subsequent calls? I don't think there is a need to manipulate them or access them directly.

It should as long as your environment is not somehow reset between calls. Use Fiddler to inspect your incoming response (with set-cookie) and outgoing requests (with cookie) and see if they match.

**Comment by mseminatore** [comment]
I made sure that my xhr instance was kept alive across all calls but it seems that cookies are discarded rather than preserved. :(

I am looking at using a different http request mechanism to see if that works. Fiddler worked on NodeJS but I wasn't able to get it to intercept UWP traffic.

**Comment by timdorr** [comment]
@mseminatore Can you move this discussion over to [the Discussions tab](https://github.com/timdorr/tesla-api/discussions)? This issue is for figuring out the new auth system and documenting it, not for specific implementation issues.

**Comment by mseminatore** [comment]
@timdorr Sorry, yes.

**Comment by ptz0n** [comment]
Thanks for sharing the new login mechanism. Work like a charm 🎉

**Comment by JSkimming** [comment]
Just adding some more information to the mix, as it may prove useful. The Tesla auth endpoint has an OpenID Connect Discovery endpoint. See below:

https://auth.tesla.com/oauth2/v3/.well-known/openid-configuration

This defines the various endpoints, supported scopes, and claims.

**Comment by denouche** [comment]
Hi @timdorr , I think the authentication process described here is not complete for some users of different market.
I opened a separated issue #299 with the things I tested and my progress.

**Comment by dustin** [comment]
step 4 states that `client_secret` is a required parameter.  I've found that sending that parameter causes an error while not including that value allows authentication to succeed.

**Comment by jasminsehic** [comment]
> Hi @timdorr , I think the authentication process described here is not complete for some users of different market.
> I opened a separated issue #299 with the things I tested and my progress.

Yeah seems that `login_hint` may be a better approach to handle all markets.

**Comment by SeBsZ** [comment]
All steps are working for me, but in m ycase I'm getting a 400 bad request and 'internal server error' when I try to refresh a token. Even in Postman this happens. Is refreshing tokens working for anyone else or is this something we still need to figure out?
![image](https://user-images.githubusercontent.com/5457531/107535282-e4b6a500-6bc0-11eb-9695-a88152244e5b.png)

**Comment by IMgoRt** [comment]
> All steps are working for me, but in m ycase I'm getting a 400 bad request and 'internal server error' when I try to refresh a token. Even in Postman this happens. Is refreshing tokens working for anyone else or is this something we still need to figure out?
> ![image](https://user-images.githubusercontent.com/5457531/107535282-e4b6a500-6bc0-11eb-9695-a88152244e5b.png)

My refresh token is way longer than that - I have a "." after that short string with a much longer 2nd part.
When I receive an HTTP 400 it was using only the first part.

**Comment by edlea** [comment]
> All steps are working for me, but in m ycase I'm getting a 400 bad request and 'internal server error' when I try to refresh a token. Even in Postman this happens. Is refreshing tokens working for anyone else or is this something we still need to figure out?

This should probably be in the Discussions Forum, but it looks like you're using a refresh token from step 4; you need to use the token from step 3

**Comment by JorenVos** [comment]
Any update on the `redirect_uri`-property? Trying to implement a UX-convenient way of logging in using https://github.com/FormidableLabs/react-native-app-auth in my RN-app; but without having the ability to set `redirect_uri` to my app scheme, it will never be possible to detect a successful login and a user ending up on a 404-error :(

**Comment by ptz0n** [comment]
@JorenVos You will have to catch the HTTP 302 response, without following it and parse the header location value.

**Comment by JorenVos** [comment]
> @JorenVos You will have to catch the HTTP 302 response, without following it and parse the header location value.

Thanks for your feedback, but that library uses some kind of blackbox implementation for iOS/Android. It might be impossible to catch that response, because it's handled in a native Safari-modal. Normally, the `redirect_uri` should be your app scheme, so the browser will redirect the user to the app with a payload. That payload contains the response data.

Sad to hear that some more manual effort will be required then.

**Comment by timdorr** [comment]
Well, Tesla's app is a React Native app, so it's totally possible. They watch for when the webview navigates to a particular set of URLs and capture the redirect data to extract the code for Step 3 and beyond. It should be possible to replicate that process in your own apps as well.

**Comment by alandtse** [comment]
> Any update on the `redirect_uri`-property?

I think we've all come to the conclusion you can't change it because it's registered by Tesla and the servers will reject the request if it's changed. Others have commented on your options though on processing it.

**Comment by JorenVos** [comment]
Hmm yes indeed, custom WebView, I see. It won't work with the native OpenID libs like [AppAuth-iOS ](https://github.com/openid/AppAuth-iOS) and [AppAuth-Android ](https://github.com/openid/AppAuth-Android), which are using `SFAuthenticationSession` and `SFSafariViewController` for iOS, and a Chrome Custom Tab implementation for Android due to security reasons. The login procedure in the main Tesla-app might not the most secure one as [there are some drawbacks using a webview for authentication](https://tools.ietf.org/html/rfc8252). Hope that Tesla changes their mind some day.

> For authorizing users in native apps, the best current practice is to perform the OAuth authorization request in an external user-agent (typically the browser) rather than an embedded user-agent (such as one implemented with web-views).

> In typical web-view-based implementations of embedded user-agents, the host application can record every keystroke entered in the login form to capture usernames and passwords, automatically submit forms  to bypass user consent, and copy session cookies and use them to perform authenticated actions as the user.

**Comment by HarmOtten** [comment]
> The Cookie header is a hard requirement for the POST request to succeed, unfortunately.

I am trying to implement the new autorisation in Node-RED.
After step1 the response header.set-cookie contains an array of 1 element. What part of this object do I need to copy to the next request?
- The array?
- The only element in the array?
- or only a part of that element?
Sorry for the newbie question, but I still get statusCode=200 in Step2 of the description.

**Comment by felickz** [comment]
> > Any update on the `redirect_uri`-property?
> 
> I think we've all come to the conclusion you can't change it because it's registered by Tesla and the servers will reject the request if it's changed. Others have commented on your options though on processing it.

This is OIDC spec to whitelist fully qualified domain in redirect uri - no wildcards.

**Comment by SG57** [comment]
Can anyone confirm this auth approach is working as of the last 24 hours?  I'm getting a 403 forbidden error 90% of the time.  That means 10% of the time I get a successful auth result then after 5 minutes the entire session expires including the refresh token forcing me to re-auth.  Note it's a 403 forbidden implying my auth is successful to a degree, perhaps its the resource/claims being requested that are limited now.

I normally sniff the official app traffic and replicate what I'm seeing but my older rooted device died and won't have a replacement until Friday.  I've inspected the official app source and am not seeing anything different than what's known already.  Worth nothing, the official app base-64 encodes its state contents, so I thought perhaps they can detect this server side and deduce it's not from the official app - still no luck.

**Comment by timdorr** [comment]
It appears Tesla's WAF is going a bit haywire at the moment. It appears you must rate limit your requests, otherwise you're going to get blocked. It's somewhere between 15-30 seconds per request (so, 2-4 rpm), so try not to request more often than that if you can.

I'm not 100% sure if varying things like your user agent or other headers will make an impact. And this may only be temporary. It's hard to know specifics without direct confirmation from Tesla.

**Comment by SearchForTheCode** [comment]
Damn Tesla is really giving us a hard time with all those stuff atm. 

Weirdly i am only doing 5 requests at all per 15-30 seconds, so it should work but i get the same 403...

**Comment by SG57** [comment]
Some research I found a similar ~80% fail rate issue opened regarding an OIDC provider behind a WAF.

https://github.com/dotnet/aspnetcore/issues/4589

I wonder if the 403 occurs due to any base-64 strings containing "--" making the WAF think it's a SQL comment and tripping the firewall.  Would explain why it works sometimes if the base-64 encoded string we generate for the code challenge, etc. Only sometimes contains a "--" sequence.  I'll look into it more but really seems only sniffing the official app traffic will tell me definitively.

**Comment by SearchForTheCode** [comment]
I have to say, i had a code_challenged with just a single "-" and it failed :( So that cant be the only thing :/ 
Also really weird that you all get a 403 , where i only get timeouts then on https://auth.tesla.com/oauth2/v3/token

And if i will get just ONE timeout on the token, all other requests before "https://auth.tesla.com/oauth2/v3/authorize" will get immediatly a timeout on the next calls. Like i was now blacklisted for 30 minutes..

**Comment by edlea** [comment]
I see timeouts on the token endpoint when running in AWS but typically things work reliably locally. 

I assumed AWS is being blocked by the WAF. Possibly your IP is being too?

**Comment by SearchForTheCode** [comment]
Not using AWS, just a local .NET program. But like i said, i copy the exact JSON body into postman of step 3 in auth mechanism and it works... in my local .NET program it does not work ... i try to find out with Fiddler what the difference is, because it dont make any sense..

**Comment by SearchForTheCode** [comment]
Here we go again, everything ran for 24-48 hours without a problem. No i am stuck to timeouts again. But on Step 1 
https://auth.tesla.com/oauth2/v3/authorize ... really frustrating.

**Comment by lannsjo** [comment]
Same for me.

**Comment by ptz0n** [comment]
@SearchForTheCode @lannsjo What does your request patterns look like?

**Comment by lannsjo** [comment]
Using this: https://github.com/tomhollander/TeslaAuth/blob/master/library/TeslaAuthHelper.cs

**Comment by SearchForTheCode** [comment]
Using completly the RestSharp Lib , everything is working again! The requests follow exactly the documentation flow, so i do nothing special. I just transfered all requests from HTTPWebRequests to Requests done via RestSharp. I dont know if they implement any WAF mechanism? ( who knows)

**Comment by lannsjo** [comment]
For me it worked once today, now its blocked again..  both locally and in AWS.

**Comment by fniewijk** [comment]
I'm building a TypeScript implementation. Yesterday I got timeouts at step 3. Today I had a timeout at step 1. I solved it by removing the `user-agent` I was sending. I tried to be a good citizen, but you seem to be punished for it. Now I'm stuck at step 2 again. Am getting a timeout.

**Comment by fniewijk** [comment]
AWS is blocked permanently in my experience. That happened around september. I believe they must have gotten a lot of brute force logging-in attempts from that platform.

**Comment by lannsjo** [comment]
Yes it was blocked in September, but started to work again around new years.

**Comment by corsair** [comment]
@fniewijk @lannsjo Running requests through AWS last year was definitely frightful, but I've been running requests across almost every endpoint through (SSO auth, token refresh, owners API) on AWS since new years without any issue. It appears there's been some overall instability within the last 24 hours, but the route of removing all headers from requests should still work for the SSO authentication depending on your codebase.

![image](https://user-images.githubusercontent.com/1847891/109134100-8cc48600-770a-11eb-97c0-87697820cec4.png)

**Comment by fniewijk** [comment]
I've started adding artificial wait times between calls. 5 seconds at the moment. This helps me now to get to the 3rd step again. I haven't gone beyond that step yet.

Maybe Tesla's servers are slow or there is some kind of race condition going on at their end.

**Comment by fniewijk** [comment]
@SearchForTheCode Could be that one library sets the `User-Agent` header automatically. I have this problem in TypeScript while using the `axios` library. I'm thinking if I should swap out that library because I can't stop it from sending certain headers.

**Comment by SearchForTheCode** [comment]
@fniewijk no , RestSharps sends by default no user-agent and i havent set any because of recommendations. But POSTMAN is using User-Agents which are enabled (and it also worked there).

**Comment by alandtse** [comment]
> Can anyone confirm this auth approach is working as of the last 24 hours? I'm getting a 403 forbidden error 75% of the time. That means 25% of the time I get a successful auth result then after 5 minutes attempting to use the refresh token the same thing - 75% of the time my refresh token request fails but then 25% it succeeds. Note it's a 403 forbidden implying my auth is successful (it's not 401 unauthorized), perhaps its the resource/claims being requested that are limited now or it's actually an issue on Tesla's side? I'll sniff the official app to see if they have the same issue.
...
The error is coming from the POST to oauth2/v3/token with the code verifier you already have and code you parse from the void callback query params (grant type = "authorization_code").

I am seeing this behavior today as well when verifying MFA codes. My implementation is a proxy where the browser passes through the JavaScript but the code to rewrite the urls to the proxy do not work on the minified JavaScript.  I suspect there's some client side check happening in the Javascript as I can get in with a different tab using the browser for the same URL when I go directly through to Tesla.

The specific response now is:
```html
<html><head>
<title>Access Denied</title>
</head><body>
<h1>Access Denied</h1>
 
You don't have permission to access "http://auth.tesla.com/oauth2/v3/authorize/mfa/verify" on this server.<p>
Reference #SNIP
</p></body>
</html>
```

**Comment by cjo2** [comment]
I'm chasing this via the JavaScript route and my "step #2" request (Obtain Authorization Code) seems to be hanging more often than not. I was able to get back an authorization code a grand total of 1 time, so it seems like my code is valid, but I'm questioning if Tesla is throttling me at this point.

Thanks for sharing all your experiences so far!

**Comment by uhthomas** [comment]
@alandtse This was a temporary issue with the Tesla WAF. Should now be resolved.

**Comment by timdorr** [comment]
Folks, Tesla's WAF going into Beast Mode doesn't have anything to do with the topic at hand. Nor do I or anyone else here (that I know of) have any means to fix the issue. I've removed anything relating to service issues on Tesla's side because there's nothing for us to do here and many of these issue comments are no longer relevant. 

**[So, please use the Discussions tab for discussing service issues.](https://github.com/timdorr/tesla-api/discussions)**

If there's a workaround to document, we can branch that out of a discussion thread into an issue (or PR!). But let's start somewhere more freeform and track it as an issue once it becomes a more concrete implementation detail that everyone needs to know.

**Comment by labdel** [comment]
Hi All,
I followed this thread and updates done in https://tesla-api.timdorr.com/api-basics/authentication (Thank you !)

I am trying this in python following code at https://github.com/enode-engineering/tesla-oauth2/blob/main/tesla.py

When executing:
resp = session.get("https://auth.tesla.com/oauth2/v3/authorize", headers=headers, params=params)

works fine when running in my local machine, ie. I am able to get the token and refresh token but if I try it in my AWS EC2 machine it never returns.

Does anyone think of anything wrong I am doing/missing ?

Thanks in advance - Leo.

**Comment by mseminatore** [comment]
Question (LMK if this is better suited to Discussion), for the MFA path what should the device name be?

**Comment by timdorr** [comment]
That doesn't appear to be part of the input. You get a set of available MFA devices from `/oauth2/v3/authorize/mfa/factors` and pick one (if there are multiple) to use for the `factor_id` on `/oauth2/v3/authorize/mfa/verify`.

**Comment by BLZB0B** [comment]
not sure if I'm not able to follow instructions - but I'm stuck on step 2 ?

instead of the location containing the code, my location field looks like:

`https://auth.tesla.com/void/callback?error=login_cancelled&error_description=User%20cancelled%20login&state=123
` 

I'm as sure as I can be that I've set the request and the body as needs be - but clearly missing something (not so) obvious.. 

Thanks

**Comment by sottileg** [comment]
@timdorr I have a working postman collection with the new authentication mechanism working and a bunch of the other endpoints too. Let me know you are interested and I can submit it as a pull request. It can be helpful for beginners in demonstrating how to make the calls described in your documentation.

**Comment by jdemeyer1** [comment]
@sottileg 
I posted my collection and environment [here](https://github.com/jdemeyer1/TeslaAPIThruPostman).

Joe

**Comment by jnicolson** [comment]
In case it helps anyone else, i've created an [Insomnia](https://insomnia.rest/) workspace.  I only created it for my own use but thought i'd share it in case anyone was interested.  The authentication flow is all in the Authentication folder and it works with MFA - not documented but i've tested and it works.  I had to install plugins to accomplish what i wanted - they are listed in the gist below.  One of the plugins generates TOTP tokens so i don't need to enter the MFA code manually each time. 

There is an environment setup but i've removed all the sensitive information - anything with a ??? you'll need to replace with your own.  Hope this helps someone!

https://gist.github.com/jnicolson/70c5d1acc1d26b0878a61fc312308b96

**Comment by hungrxyz** [comment]
> @timdorr I have a working postman collection with the new authentication mechanism working and a bunch of the other endpoints too. Let me know you are interested and I can submit it as a pull request. It can be helpful for beginners in demonstrating how to make the calls described in your documentation.

@sottileg  I'm trying to use the API and a working collection would help if you can share it with us please.

**Comment by lannsjo** [comment]
For me the authentication has stopped worked. Isnt there anyone here that can provide the authentication endpoints as a service. Would happily pay for this to have a working solution that doesnt need to be fixed every month..

**Comment by Evogelpohl** [comment]
Same same. I've had a PowerShell based script running in Azure Automation (South Central region), per hour, get vehicle_data (save json to blob storage). Worked just fine up to 4/28/21 @ 11AM. Then stopped working. "The remote server returned an error: (401) Unauthorized." call the Auth function w/in the script.

Same code works fine in VSCode on my desktop as of today. Figured Azure was blacklisted by CDN/FW of some sort.

**Comment by jdemeyer1** [comment]
The /oauth2/v3/ token is after the Obtain Authorization Code call.  The Obtain Authorization Code call is failing for me.

After I POST the Obtain Authorization request, I received three GET responses:
One GET to auth.telsa.com/oatuh2/v3/authorize, one GET to auth.tesla.com/void/callback (I think I have to use this one to extract "code"), and one GET that is 404.html.

I attempt to get a referer from the request object but I think the last GET (with the 404.html) is failing my code.  That is, I need to get a referer from the second GET rather than the third GET.  Not sure how to do that just yet.

**Comment by carlosonunez** [comment]
I'm getting 403's as well at the `/authorize` endpoint while trying to generate access tokens from my Ruby script [here](https://github.com/carlosonunez/tesla-oauth2-token-bot).

**Comment by JorenVos** [comment]
@carlosonunez It's also returning 403's when trying to login in the official Tesla-app, so I guess it's broken at their side.

**Comment by jnicolson** [comment]
I noticed when you try to login via the Web, instead of /oauth2/v3/ it's /oauth2/v1/.  Does anyone remember whether the web auth was always v1?

I tried to use /v1/ with the ownerapi (i.e. pretending to be the app) and I was able to get through the whole authorisation flow (i.e. I have my access token and can call the API).  Doesn't help me log back into the app on my phone though (I logged out to test what it was going to do).

**Comment by jnicolson** [comment]
.. and now that i've written that, the /v3/ seems to be working again.  Web still points to /v1/ though, and they both seem to work.

**Comment by pickenheim** [comment]
Since May 8th late afternoon EST /v3/ had 403's, just tried /v1/ and it worked nominally. Tesla vehicle app had message on bottom of screen that I have never seen before (recently May 9th), usually it is "last updated at" when it can't update, this was "down for maintenance check back" something like that and went away a few minutes later.

**Comment by carlosonunez** [comment]
Thank goodness! I’ll try changing my paths when I get home.

**Comment by pickenheim** [comment]
It gets all the way through the authorization process but the final response returning the tokens on /v1/ is Internal Server Error 500, so I'm going to chill, give it some time, make the code have a way to wait and try again after a timeout. I am making something to estimate demand standby power usage for solar, there are surcharges to avoid, turn on Powerwalls for a short period, turn off Tesla if charging first, avoid grid load for expensive peaks, works well just needs refinement.

**Comment by chrm-software** [comment]
Changing /v3/ to /v1/ in auth process works for me too. No HTTP 500 here. Please check your headers.

**Comment by carlosonunez** [comment]
I got HTTP 500 invalid code_verifier when I did this. Do I need to change the challenge generation mechanism as well?

**Comment by davidhodge** [comment]
FYI:

I’m seeing some data with Nikola (an app I make) that suggests they’ve been dealing with widespread partial outages of their vehicle api since Friday. Findings are somewhat consistent with them deploying another change early this AM causing auth issues.

David

**Comment by carlosonunez** [comment]
DUDE!!!!! Thank you for making Nikola! Unfortunately I made the mistake of logging out 🤣

I’m getting similar reports from Redditors in a thread I made about this. Someone had trouble taking delivery of their car because of this.

**Comment by chrm-software** [comment]
I didn't change anything else in my code, just the version string in the URI.

---

> I got HTTP 500 invalid code_verifier when I did this. Do I need to change the challenge generation mechanism as well?
> […](#)
> On May 9, 2021, 11:16 -0500, Christoph , wrote: Changing /v3/ to /v1/ in auth process works for me too. No HTTP 500 here. Please check your headers. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

**Comment by pickenheim** [comment]
OKAY v3 just came back online for me at least, thank the Maker!

**Comment by carlosonunez** [comment]
Looks like sign-in is working again on /v3. Can anyone confirm?

**Comment by jnicolson** [comment]
> Looks like sign-in is working again on /v3. Can anyone confirm?
> […](#)
> On May 9, 2021, 12:47 -0500, pickenheim ***@***.***>, wrote: > Changing /v3/ to /v1/ in auth process works for me too. No HTTP 500 here. Please check your headers. I still get this error in the final step to get tokens. I reviewed the header, JSON all that, had worked for weeks as /v3/. Will update if I learn anything else, expecting it to come online at some point soon. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub, or unsubscribe.

v3 back up for me too

> > Changing /v3/ to /v1/ in auth process works for me too. No HTTP 500 here. Please check your headers.
> 
> I still get this error in the final step to get tokens. I reviewed the header, JSON all that, had worked for weeks as /v3/. Will update if I learn anything else, expecting it to come online at some point soon.

I actually had the same issue - my testing setup in Insomnia which worked for v3 didn't work for v1.  However when I read the docs again it turns out i was doing the code_verifier and code_challenge completely wrong but v3 was still accepting it (basically i was creating code_ challenge just by taking the base64 of the code_verifier without a SHA256 sum and this was being accepted).  v1 is more strict by the looks.

I'd check that you're generating code_ challenge correctly and then sending the code_verifier used to create code_challenge.

**Comment by carlosonunez** [comment]
Okay, I'm still getting 403s when using HTTParty, Faraday, or Net::HTTP. cURL does not have this problem. I get this even when their User-Agent headers are set to match `cURLs` (cURL is working).

Something weird is happening.

**Comment by spoonwzd** [comment]
I'm still getting 403/forbidden when trying to authenticate against /v3 from my home internet connection, even after my IP address has changed. The Tesla app works fine from my home connection and the same code works fine from other IP addresses .

Can anyone shine a light on what this might be? I've about run out of hair...

**Comment by lannsjo** [comment]
> Its something about the environment - move the same code to another location / internet connection and it works fine. But I've changed my IP address - what else could it be?

This is unfortunately how we are doing it at the moment as well. It works for a month, then it suddenly stops working, move it to another hosting provider and it all works again.
